File: crawler/crawler.js
require('dotenv').config();
const { chromium } = require('playwright');
const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');
const axios = require('axios');

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_ANON_KEY
);

const SCREENSHOT_DIR = path.join(__dirname, 'screenshots');
if (!fs.existsSync(SCREENSHOT_DIR)) fs.mkdirSync(SCREENSHOT_DIR);

// ================================================
// Claude Vision으로 가격 추출
// ================================================
async function extractPriceWithClaude(screenshotPath, modelName, channel) {
  const imageData = fs.readFileSync(screenshotPath);
  const base64Image = imageData.toString('base64');

  const channelNames = { coupang: '쿠팡', gmarket: 'G마켓', himart: '하이마트' };

  try {
    const response = await axios.post(
      'https://api.anthropic.com/v1/messages',
      {
        model: 'claude-sonnet-4-6',
        max_tokens: 500,
        messages: [
          {
            role: 'user',
            content: [
              { type: 'image', source: { type: 'base64', media_type: 'image/png', data: base64Image } },
              {
                type: 'text',
                text: `이 이미지는 ${channelNames[channel]}에서 "${modelName}" 모델을 검색한 결과입니다.

가장 낮은 판매가격(최저가)을 만원 단위 정수로만 알려주세요.
예: 278 (278만원인 경우)

숫자만 답하세요. 가격을 찾을 수 없으면 0으로 답하세요.`
              }
            ]
          }
        ]
      },
      {
        headers: {
          'x-api-key': process.env.ANTHROPIC_API_KEY,
          'anthropic-version': '2023-06-01',
          'content-type': 'application/json'
        }
      }
    );

    const text = response.data.content[0].text.trim();
    const price = parseInt(text.replace(/[^0-9]/g, ''));
    return isNaN(price) ? 0 : price;
  } catch (err) {
    console.error('Claude Vision 오류:', err.message);
    return 0;
  }
}

// ================================================
// 채널별 크롤링 함수
// ================================================
async function crawlCoupang(page, modelName) {
  const searchUrl = `https://www.coupang.com/np/search?q=${encodeURIComponent(modelName)}&channel=user`;
  await page.goto(searchUrl, { waitUntil: 'networkidle', timeout: 30000 });
  await page.waitForTimeout(2000);
  return page;
}

async function crawlGmarket(page, modelName) {
  const searchUrl = `https://browse.gmarket.co.kr/search?keyword=${encodeURIComponent(modelName)}`;
  await page.goto(searchUrl, { waitUntil: 'networkidle', timeout: 30000 });
  await page.waitForTimeout(2000);
  return page;
}

async function crawlHimart(page, modelName) {
  const searchUrl = `https://www.himart.co.kr/search/list?keyword=${encodeURIComponent(modelName)}`;
  await page.goto(searchUrl, { waitUntil: 'networkidle', timeout: 30000 });
  await page.waitForTimeout(2000);
  return page;
}

const crawlFunctions = { coupang: crawlCoupang, gmarket: crawlGmarket, himart: crawlHimart };

// ================================================
// 메인 크롤링 실행 (새 스키마: product_groups + channel_models)
// ================================================
async function runCrawler() {
  console.log(`\n🚀 크롤링 시작: ${new Date().toLocaleString('ko-KR')}`);
  let savedCount = 0;

  const { data: groups, error } = await supabase
    .from('product_groups')
    .select(`
      id, brand, ata_price,
      categories(name),
      channel_models(id, channel, model_name, search_url)
    `)
    .eq('is_active', true);

  if (error) {
    console.error('모델 조회 오류:', error.message);
    return { count: 0 };
  }

  if (!groups || groups.length === 0) {
    console.log('⚠️ 등록된 모델이 없습니다. 대시보드에서 모델을 추가해주세요.');
    return { count: 0 };
  }

  // channel_models를 평탄화
  const items = [];
  groups.forEach(g => {
    (g.channel_models || []).forEach(cm => {
      items.push({
        channel_model_id: cm.id,
        channel: cm.channel,
        model_name: cm.model_name,
        ata_price: g.ata_price,
        brand: g.brand,
        category_name: g.categories?.name
      });
    });
  });

  if (items.length === 0) {
    console.log('⚠️ 채널별 모델명이 등록된 항목이 없습니다.');
    return { count: 0 };
  }

  const browser = await chromium.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });

  const context = await browser.newContext({
    viewport: { width: 1280, height: 900 },
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
  });

  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);

  for (const item of items) {
    console.log(`\n📦 ${item.category_name} | ${item.brand} | ${item.channel} | ${item.model_name}`);
    const page = await context.newPage();

    try {
      console.log(`  ⏳ 크롤링 중...`);
      await crawlFunctions[item.channel](page, item.model_name);

      const screenshotName = `${timestamp}_${item.model_name}_${item.channel}.png`;
      const screenshotPath = path.join(SCREENSHOT_DIR, screenshotName);
      await page.screenshot({ path: screenshotPath, fullPage: false });

      const price = await extractPriceWithClaude(screenshotPath, item.model_name, item.channel);
      const ataDiff = item.ata_price ? price - item.ata_price : null;

      const { error: insertError } = await supabase
        .from('price_history')
        .insert({
          channel_model_id: item.channel_model_id,
          price: price,
          screenshot_url: screenshotName,
          ata_diff: ataDiff,
          crawled_at: new Date().toISOString()
        });

      if (insertError) {
        console.error(`  ❌ DB 저장 오류:`, insertError.message);
      } else {
        savedCount++;
        console.log(`  ✅ ${price}만원 (ATA ${ataDiff !== null ? (ataDiff >= 0 ? '+' : '') + ataDiff : '-'}만원)`);
      }
    } catch (err) {
      console.error(`  ❌ 오류:`, err.message);
    } finally {
      await page.close();
    }

    await new Promise(r => setTimeout(r, 3000));
  }

  await browser.close();
  console.log(`\n✅ 크롤링 완료: ${new Date().toLocaleString('ko-KR')} (${savedCount}건 저장)\n`);
  return { count: savedCount };
}

// 직접 실행 시 (node crawler.js)
if (require.main === module) {
  runCrawler().catch(console.error);
}

module.exports = { runCrawler };

File: crawler.py
import os
import time
import random
import base64
import json
from datetime import datetime
from DrissionPage import ChromiumPage, ChromiumOptions
from supabase import create_client
import requests

# ================================================
# 설정
# ================================================
SUPABASE_URL = "https://ycopchmcvjjbcfopuywy.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Inljb3BjaG1jdmpqYmNmb3B1eXd5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3ODI1NjI3NzQsImV4cCI6MjA5ODEzODc3NH0.maP1adRd7tSuChZw5TtiMEyXPl8vm8ZosGI-ZyMy0QA"
ANTHROPIC_API_KEY = os.environ.get("ANTHROPIC_API_KEY", "여기에 키 입력")

SCREENSHOT_DIR = "screenshots"
os.makedirs(SCREENSHOT_DIR, exist_ok=True)

supabase = create_client(SUPABASE_URL, SUPABASE_KEY)

CHANNEL_NAMES = {"coupang": "쿠팡", "gmarket": "G마켓", "himart": "하이마트"}

# ================================================
# Claude Vision으로 가격 추출
# ================================================
def extract_price_with_claude(screenshot_path, model_name, channel):
    with open(screenshot_path, "rb") as f:
        image_data = base64.b64encode(f.read()).decode("utf-8")

    try:
        response = requests.post(
            "https://api.anthropic.com/v1/messages",
            headers={
                "x-api-key": ANTHROPIC_API_KEY,
                "anthropic-version": "2023-06-01",
                "content-type": "application/json"
            },
            json={
                "model": "claude-sonnet-4-6",
                "max_tokens": 500,
                "messages": [
                    {
                        "role": "user",
                        "content": [
                            {
                                "type": "image",
                                "source": {
                                    "type": "base64",
                                    "media_type": "image/png",
                                    "data": image_data
                                }
                            },
                            {
                                "type": "text",
                                "text": f"이 이미지는 {CHANNEL_NAMES[channel]}에서 '{model_name}' 모델의 상품 페이지입니다.\n\n판매가격(최저가)을 원 단위 정수로만 알려주세요.\n예: 3313610 (3,313,610원인 경우)\n\n쉼표, 원, 공백 없이 숫자만 답하세요. 가격을 찾을 수 없으면 0으로 답하세요."
                            }
                        ]
                    }
                ]
            }
        )
        text = response.json()["content"][0]["text"].strip()
        price = int("".join(filter(str.isdigit, text)))
        return price
    except Exception as e:
        print(f"  Claude Vision 오류: {e}")
        return 0

# ================================================
# 채널별 URL 생성
# ================================================
def get_url(item):
    if item.get("search_url"):
        return item["search_url"]
    
    model = item["model_name"]
    channel = item["channel"]
    
    if channel == "coupang":
        return f"https://www.coupang.com/np/search?q={model}&channel=user"
    elif channel == "gmarket":
        return f"https://browse.gmarket.co.kr/search?keyword={model}"
    elif channel == "himart":
        return f"https://www.himart.co.kr/search/list?keyword={model}"

# ================================================
# 메인 크롤러
# ================================================
def run_crawler():
    print(f"\n🚀 크롤링 시작: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    saved_count = 0

    # DB에서 모델 가져오기
    result = supabase.table("product_groups")\
        .select("id, brand, ata_price, categories(name), channel_models(id, channel, model_name, search_url)")\
        .eq("is_active", True)\
        .execute()

    groups = result.data
    if not groups:
        print("⚠️ 등록된 모델이 없습니다.")
        return

    # 평탄화
    items = []
    for g in groups:
        for cm in g.get("channel_models", []):
            items.append({
                "channel_model_id": cm["id"],
                "channel": cm["channel"],
                "model_name": cm["model_name"],
                "search_url": cm.get("search_url"),
                "ata_price": g.get("ata_price"),
                "brand": g["brand"],
                "category_name": g.get("categories", {}).get("name", "")
            })

    if not items:
        print("⚠️ 채널별 모델이 없습니다.")
        return

    # DrissionPage 브라우저 설정
    options = ChromiumOptions()
    options.headless(False)  # 창 띄우기 (봇 감지 우회)
    options.set_argument("--no-sandbox")
    options.set_argument("--disable-blink-features=AutomationControlled")
    options.set_user_agent(
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"
    )

    page = ChromiumPage(options)
    timestamp = datetime.now().strftime("%Y-%m-%dT%H-%M-%S")

    for item in items:
        print(f"\n📦 {item['category_name']} | {item['brand']} | {item['channel']} | {item['model_name']}")

        try:
            url = get_url(item)
            print(f"  ⏳ 접속 중: {url[:60]}...")
            page.get(url)

            # 랜덤 딜레이 (봇 감지 우회)
            delay = random.uniform(3, 6)
            time.sleep(delay)

            # 스크린샷
            screenshot_name = f"{timestamp}_{item['model_name']}_{item['channel']}.png"
            screenshot_path = os.path.join(SCREENSHOT_DIR, screenshot_name)
            page.get_screenshot(path=SCREENSHOT_DIR, name=screenshot_name)

            # Claude Vision으로 가격 추출
            price = extract_price_with_claude(screenshot_path, item["model_name"], item["channel"])

            # ATA 차이 계산 (원단위)
            ata_price_won = item["ata_price"] * 10000 if item["ata_price"] else None
            ata_diff = price - ata_price_won if ata_price_won and price else None

            # DB 저장
            supabase.table("price_history").insert({
                "channel_model_id": item["channel_model_id"],
                "price": price,
                "screenshot_url": screenshot_name,
                "ata_diff": ata_diff,
                "crawled_at": datetime.now().isoformat()
            }).execute()

            saved_count += 1
            price_man = round(price / 10000) if price else 0
            print(f"  ✅ {price_man}만원 ({price:,}원)")

        except Exception as e:
            print(f"  ❌ 오류: {e}")

        # 채널 간 랜덤 딜레이
        time.sleep(random.uniform(3, 7))

    page.quit()
    print(f"\n✅ 완료: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')} ({saved_count}건 저장)\n")

if __name__ == "__main__":
    run_crawler()
